reflect/src/main/java/dagger/reflect/ReflectiveComponentParser.java

```diff
@@ -0,0 +1,53 @@
+package dagger.reflect;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import static dagger.reflect.Reflection.findQualifier;
+
+final class ReflectiveComponentParser {
+  private static final Binding<?>[] NO_BINDINGS = new Binding<?>[0];
+
+  static void parse(Class<?> moduleClass, Object instance,
+      BindingGraph.Builder graphBuilder) {
+    Set<Class<?>> seen = new LinkedHashSet<>();
+    Deque<Class<?>> queue = new ArrayDeque<>();
+    queue.add(moduleClass);
```

Comment:
Should the hierarchy walking be extracted and shared between the two parsers in the future?

Reply:
Sure. Makes sense to de-duplicate once their both merged.